### ISSUE TITLE
MapEncoding for ETH set

### ISSUE THREAD

**Issue by TsuTikgiau** [StanfordASL/Trajectron-plus-plus#14]
Hello!
I'd like to check the performance of map encoding on ETH. I notice that in your paper you do some experiments on the ETH - University scene with the map encoding to show the advantage. Does this repository also support this experiment? I checked the processed ETH - University dataset and noticed that scene.map in this dataset is None. Do I miss something? Thank you for your help!

**Comment by BorisIvanovic** [comment]
Hello!
You didn't miss anything, this is a good question. 

This isn't currently implemented in the latest code because the provided ETH maps are image-based (the ETH maps are literally jpeg or png images, I forget which in particular) whereas the nuScenes maps are geometric/rasterized. In this codebase, we chose to focus on the autonomous driving use case and thus on supporting geometric/rasterized maps. 

Not all is lost, this is actually the single reason I've kept around the old version of this codebase in the [old_pedestrians](https://github.com/StanfordASL/Trajectron-plus-plus/tree/old_pedestrians) branch, so you can see what it looked like and how it took in image-based maps.

To visualize the same experiment you'll have to port over code from the old version into this repo. I'm sorry that we didn't do this ourselves.

**Comment by TsuTikgiau** [comment]
Thanks for your answering! I checked the old_pedestrians branch, but I still didn't find any map information there (config files of pretrained models under /data/models don't use map encoder, only txt files in /data/raw, and /data/process_data.py doesn't take maps into account). Do you know how can I run some experiments and evaluate for the eth map encoding using the old branch? Thanks!

**Comment by BorisIvanovic** [comment]
You can see the map encoding [here](https://github.com/StanfordASL/Trajectron-plus-plus/blob/old_pedestrians/code/model/node_model.py#L452) and the CNN that was used [here](https://github.com/StanfordASL/Trajectron-plus-plus/blob/old_pedestrians/code/model/components/map_encoder.py#L6).

As for obtaining the actual maps, they can be found in the original ETH dataset files: https://data.vision.ee.ethz.ch/cvl/aem/ewap_dataset_full.tgz 

As for including the maps in the Scene object, you would instantiate a [Map](https://github.com/StanfordASL/Trajectron-plus-plus/blob/8a1fd51801e6efab6af9a48909d285d9fb85b7d5/code/data/data_structures.py#L60) object with the loaded map + homography matrix and then add the Map object as an argument to the Scene object instantiation in the [process_data.py](https://github.com/StanfordASL/Trajectron-plus-plus/blob/8a1fd51801e6efab6af9a48909d285d9fb85b7d5/data/process_data.py#L80) script. This should get you most of the way to training your own model with map encoding.

Conversely, you could also copy that Map object and associated code to the new codebase and do the same with the new codebase's map encoding (I recommend training a new model in the new codebase rather than the old one, training should be much faster because of our use of PyTorch's parallel data loading).

**Comment by TsuTikgiau** [comment]
Thank you for your kindly reply! I have a question about the homography matrix. I notice that the homography matrix in the original dataset file 'H.txt' brings the point from image to world coordinates. Can I directly use the inverse of this matrix to instantiate the Map object here? Besides, I notice a main difference between the old version and the new version of Map object:
 [https://github.com/StanfordASL/Trajectron-plus-plus/blob/8a1fd51801e6efab6af9a48909d285d9fb85b7d5/code/data/data_structures.py#L73](url) 
[https://github.com/StanfordASL/Trajectron-plus-plus/blob/master/trajectron/environment/map.py#L177](url). The old version use the np.fliplr function when brings the point to the map coordinate system. Why do we do this, and should I keep this operation when I using the new codebase? Thanks!

**Comment by TsuTikgiau** [comment]
I tried to implement the ImageMap class [here](https://github.com/StanfordASL/Trajectron-plus-plus/blob/master/trajectron/environment/map.py#L183) like this
`
class ImageMap(GeometricMap): 

    def to_map_points(self, scene_pts):
        org_shape = None
        if len(scene_pts.shape) > 2:
            org_shape = scene_pts.shape
            scene_pts = scene_pts.reshape((-1, 2))
        N, dims = scene_pts.shape
        points_with_one = np.ones((dims + 1, N))
        points_with_one[:dims] = scene_pts.T
        map_points = np.fliplr((self.homography @ points_with_one).T[..., :dims])
        if org_shape is not None:
            map_points = map_points.reshape(org_shape)
        return map_points
`
Basically it just replaces the to_map_points in GeometricMap class using the old version. And using the inverse matrix of H.txt as the homography and the map data to initialize the new ImageMap object and send it to scene. But to_map_points doesn't work correctly in this case. It still output some locations outside the maps. Do you have any suggestions? Thanks in advance!

**Comment by BorisIvanovic** [comment]
The `fliplr` usage is one of the main reasons we hated working with image-based maps 😛 It's because (x, y) indexes geometric maps just fine, but for images you start from the top left and the height is usually the first coordinate so (y, x) ends up being the coordinates used. This may require a bit of trial and error, but I remember that was a pain point for us to figure out.

As for why it doesn't work, when you loaded the homography matrix did you normalize it? This is how we loaded the homography matrix (we did indeed take the inverse as you mentioned):
```
homography_matrix = np.loadtxt('H.txt')
homography_matrix /= homography_matrix[2, 2]
homography_matrix = np.linalg.inv(homography_matrix)
```

**Comment by TsuTikgiau** [comment]
Thank you for your help! Indeed, I didn't normalize the homography matrix before. I tried normalization and notice that the code can run on ETH set. For Hotel set I still have the problem where some points are outside the map. I checked the process_data.py file and noticed that there are also data from other [datasets](https://github.com/StanfordASL/Trajectron-plus-plus/tree/master/experiments/pedestrians/raw/hotel/train)  in raw training and validation data files. Does this means that some data in the processed pkl file is not from hotel and hence cannot be compatible with the hotel map. Am I correct? If yes, why shall we also use other data to train on the hotel set? Thanks for your help!

**Comment by BorisIvanovic** [comment]
You are correct, and this is actually on purpose. The reason for this is that a leave-one-out strategy is employed for training, validation, and testing on these pedestrian datasets.

What this means is that the model is trained on all the [other pedestrian datasets](https://github.com/StanfordASL/Trajectron-plus-plus/tree/master/experiments/pedestrians/raw/hotel/train), validated on [different subsets of those datasets](https://github.com/StanfordASL/Trajectron-plus-plus/tree/master/experiments/pedestrians/raw/hotel/val), and then tested on the [full Hotel set](https://github.com/StanfordASL/Trajectron-plus-plus/tree/master/experiments/pedestrians/raw/hotel/test). 

A reason for why one may want to do this is that it truly tests the model on unseen data and shows the model's capability to generalize to different scene layouts and groups of pedestrians.

**Comment by TsuTikgiau** [comment]
Got it, thank you for your kindly help!
